arangod/Pregel/Actor/include/Actor/Actor.h

```diff
@@ -0,0 +1,248 @@
+        std::unique_ptr<typename Config::State> initialState,
+        std::size_t batchSize)
+      : pid(pid),
+        runtime(runtime),
+        state(std::move(initialState)),
+        batchSize(batchSize) {}
+
+  ~Actor() = default;
+
+  auto typeName() -> std::string_view override { return Config::typeName(); };
+
+  void process(ActorPID sender,
+               std::unique_ptr<MessagePayloadBase> msg) override {
+    if (auto* m =
+            dynamic_cast<MessagePayload<typename Config::Message>*>(msg.get());
+        m != nullptr) {
+      push(sender, std::move(m->payload));
+    } else if (auto* n = dynamic_cast<MessagePayload<ActorError>*>(msg.get());
+               n != nullptr) {
+      push(sender, std::move(n->payload));
+    } else {
+      runtime->dispatch(
+          pid, sender,
+          ActorError{UnknownMessage{.sender = sender, .receiver = pid}});
+    }
+  }
+
+  void process(ActorPID sender, velocypack::SharedSlice msg) override {
+    if (auto m =
+            inspection::deserializeWithErrorT<typename Config::Message>(msg);
+        m.ok()) {
+      push(sender, std::move(m.get()));
+    } else if (auto n = inspection::deserializeWithErrorT<ActorError>(msg);
+               n.ok()) {
+      push(sender, std::move(n.get()));
+    } else {
+      auto error =
+          ActorError{UnknownMessage{.sender = sender, .receiver = pid}};
+      auto payload = inspection::serializeWithErrorT(error);
+      if (payload.ok()) {
+        runtime->dispatch(pid, sender, payload.get());
+      } else {
+        fmt::print("Error serializing UnknownMessage");
+        std::abort();
+      }
+    }
+  }
+
+  auto finish() -> void override { finished.store(true); }
+  auto finishedAndIdle() -> bool override {
+    return finished.load() and idle.load();
+  }
+
+  auto serialize() -> velocypack::SharedSlice override {
+    auto res = inspection::serializeWithErrorT(*this);
+    if (not res.ok()) {
+      std::abort();
+    }
+    return res.get();
+  }
+
+ private:
+  void push(ActorPID sender, typename Config::Message&& msg) {
+    pushToQueueAndKick(std::make_unique<InternalMessage>(
+        sender,
+        std::make_unique<MessageOrError<typename Config::Message>>(msg)));
+  }
+  void push(ActorPID sender, ActorError&& msg) {
+    pushToQueueAndKick(std::make_unique<InternalMessage>(
+        sender,
+        std::make_unique<MessageOrError<typename Config::Message>>(msg)));
+  }
+
+  void kick() {
+    // Make sure that *someone* works here
+    (*runtime->scheduler)([self = this->weak_from_this()]() {
+      auto me = self.lock();
+      if (me != nullptr) {
+        me->work();
+      }
+    });
+  }
+
+  void work() {
+    if (finished.load() == true) {
+      return;
+    }
+    auto i = batchSize;
+
+    while (auto msg = inbox.pop()) {
+      state = std::visit(
+          typename Config::template Handler<Runtime>{
+              {pid, msg->sender, std::move(state), runtime}},
+          msg->payload->item);
```

Comment:
Why do you need other part of payload?

What about using two vector<variant> queue instead of allocate every message?

Reply:
What do you mean with "other part"?

Reply:
If I don't misread you use only `->item` from `payload`, why we need to make payload separate struct if used only single field ?

Reply:
Initially I wanted the concatenator in Message.h to inherit from std::variant<Args0..., Args1...> and don't have any item. But here we call visit on a variant and unfortunately gcc 11.2 - which we use on Jenkins - has a bug: I cannot call visit on an inherited variant (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90943) - it is only fixed in gcc 11.3. Therefore this is a workaround.

Reply:
Ok